ext/dom/xml_serializer.c

```diff
@@ -561,7 +591,26 @@ static int dom_xml_serialize_attribute_node_value(xmlOutputBufferPtr out, xmlAtt
 	return xmlOutputBufferWriteLit(out, "\"");
 }
 
-/* Spec says to do nothing, but that's inconsistent/wrong, see https://github.com/w3c/DOM-Parsing/issues/28 */
+/* These steps are from the attribute serialization algorithm's well-formed checks. */
+static zend_always_inline int dom_xml_check_xmlns_attribute_requirements(const xmlAttr *attr)
+{
+	const xmlChar *attr_value = dom_get_attribute_value(attr);
+
+	/* 3.5.2.2. If the require well-formed flag is set and the value of attr's value attribute matches the XMLNS namespace, then throw an exception */
+	if (strcmp((const char *) attr_value, DOM_XMLNS_NS_URI) == 0) {
+		return -1;
+	}
+
+	/* 3.5.2.3. If the require well-formed flag is set and the value of attr's value attribute is the empty string */
+	if (*attr_value == '\0') {
+		return -1;
+	}
+
+	return 0;
+}
```

Comment:
The thing is that the return values of these serialization functions must be compatible with the libxml interface, which uses an int with negative error codes. Changing this function to return a bool is possible, but would require changes at the callsites to transform the return value which I rather not do because it can be error-prone.

Reply:
Right, please add a comment above the function explaining this then :)